### ISSUE TITLE
When macOS  sleep and reopen problem

### ISSUE THREAD

**Issue by spacexnasa** [AdguardTeam/AdguardForMac#540]
hello   when macOS  sleep and reopen, adguard can not work,need  disable and enable button,then adguard work again.

os:macos 10.14.5
safari:12.1.1
adguard:  2.0.6

**Comment by michael1900** [comment]
I have not this problem, can not confirm.

**Comment by spacexnasa** [comment]
> I have not this problem, can not confirm.

it's very strange

**Comment by michael1900** [comment]
I am using last nightly. Try.

**Comment by ameshkov** [comment]
To troubleshoot this issue, we need to get the app logs.

Here's what we need you to do:

1. Menu -> Advanced -> Logging level -> Debug
2. Reproduce the issue (and remember the exact time when it happened).
3. Menu -> Advanced -> Export Logs and System Info...
4. Send the archive to `apple@adguard.com` and mention this issue number in the subject

**Comment by planet0** [comment]
Might be same issue as #489.

**Comment by zzebrum** [comment]
@spacexnasa hi there! Is the problem still persist with the latest beta/nightly?

**Comment by iDjay** [comment]
had the issue just a few minutes ago, AdGuard was running in the background and filters were up-to-date but it didn't filter anything. My workaround was too completely close and re-open Safari, after that everything was fine again.

**Comment by zzebrum** [comment]
@spacexnasa @iDjay ok guys, we need to get log files as [described above](https://github.com/AdguardTeam/AdguardForMac/issues/540#issuecomment-504915211)

**Comment by iDjay** [comment]
Can confirm again. 
1. wake up MacBook
2. launch Safari quickly
3. AdGuard doesn't work

[adguard_logs_20190928084300.zip](https://github.com/AdguardTeam/AdguardForMac/files/3665241/adguard_logs_20190928084300.zip)

**Comment by nexttoyou** [comment]
This has most probably the same origin as #599 and is obviously related to the slow wakeup of some network interfaces (might be everything except Wi-Fi). I tried solutions like [SleepWatcher](https://www.bernhard-baehr.de) and the command **restart_adguard** inside of the AdGuard app wrapper but failed.

Is there any reliable method to restart AdGuard via the command line?

Thanks for fixing this ASAP anyway!

**Comment by ameshkov** [comment]
@nexttoyou yeah, #599 must be the same issue. This also gives me the idea of how to reproduce this issue easily:

1. Disable wi-fi router
2. Connect to your phone's hotspot 
3. Sleep
4. Disable hotspot, enable wi-fi router
5. Wake
6. Filtering should stop working on some websites

**Comment by ameshkov** [comment]
> Is there any reliable method to restart AdGuard via the command line?

Hm, not really. You can try running `adguard-tcpkill` from `/Applications/Adguard.app/Contents/Resources` to force close active connections, but you'll have to pass PIDs of the browser processes to it.

**Comment by Chinaski1** [comment]
One more user faced this issue: **2369244**
17/12/2019 | 10.14 AM

**Comment by akmLaVx** [comment]
User 2369244 would like an update on the issue.

**Comment by zzebrum** [comment]
@akmLaVx the issue is assigned on current milestone. Devs will look into it soon.

**Comment by vozersky** [comment]
@Chinaski1 Can anyone confirm it still happens? 
I have to move this issue to the next milestone for now. We were not able to reproduce and debug it.

**Comment by Chinaski1** [comment]
There have been no more reports like this.

**Comment by iDjay** [comment]
Nope seems to be fixed.

**Comment by nexttoyou** [comment]
Still happens here with 2.3.2.710 on macOS 10.14.6.

**Comment by vozersky** [comment]
@nexttoyou thanks, reopened, but I will have to move it to the next milestone since we were not able to debug it so far

**Comment by iDjay** [comment]
Maybe it‘s only a Mojave issue? I‘m on Catalina and having no issues.

**Comment by nexttoyou** [comment]
Thanks. It might be resolved with Catalina but, unfortunately, changing to Catalina would give me just too many other headaches. What I already can say, however, is that one of the origins of the problem is without any doubt everything that provokes an unwanted network connection reset like described [here](https://github.com/lionheart/openradar-mirror/issues/20271), [here](https://discussions.apple.com/thread/8497345) or [here](https://discussions.apple.com/thread/250092008). AdGuard just does not seem to take well any sudden network change/wake or whatsoever and needs its time to be ready as opposed to e.g. Little Snitch.

**Comment by Ribero** [comment]
Hi, I'm facing something that looks like similar to this issue. ADG is running in background but filtering is no more working. Killing ADG process then restarting it makes it work until the next time.

The context where the issue raise is:

- macOS Mojave, Safari
- multiple user logged on the mac and each are running ADG
- user switching
- sleep / wake 

When only one user is logged on I found that the issue occurs less often.

I still need to capture debug logs when this issue is happening. I will send it to support accordingly.

**Comment by nexttoyou** [comment]
This became one of the most annoying bugs on my systems as working around it just seems to be impossible. The AdGuard filtering cannot be re-triggered by whatever script and if it stops it does not communicate it at all.

As most of the typical Mac users nowadays seem to use a Wi-Fi connection for everything it already looks for Joe User like the bug does not even exist as Wi-Fi is able to always wake up fast enough for AdGuard to be activated. But if you’re using Ethernet with link aggregation and/or dual stack IPv4/IPv6 you’re in heavy trouble.

This is one side of the medal. But very often AdGuard also suddenly stops without cause and all the ads reappear. Then it is impossible to know why and what to do against it because the reason remains unknown.

Last but not least the solution in favour to circumvent this bug is the use of Apple’s future “NetworkExtension” API (NE) instead of the current kernel extension. Unfortunately, this would mean a forced update to whatever horrible following macOS Cata(lina|stropha). The AdGuard developers might excuse my pessimism but this is not at all promising.

**Comment by Ribero** [comment]
Maybe ADG needs a sort of Watchdog process to monitor if agent is still effectively filtering ads (from a test page for example) and then kill/restart agents if not.

Killing all ADG* processes manually then relaunching the application did the trick for me, but it's clearly not something I want to use in production everyday.

**Comment by nexttoyou** [comment]
Out of despair I finally did what currently seems to work. I installed [SleepWatcher 2.2.2](https://www.bernhard-baehr.de) which executes at every wake of the Mac a small command line killing and restarting AdGuard like this:

`ps alx | egrep /Applications/Adguard.app/Contents/MacOS/Adguard | awk '{print $2}' | sudo xargs kill; open -a Adguard`

This is by no means a clean/perfect solution but does until now what it should: waking up AdGuard reliably. I also deactivated IPv6 in the hope that it does not confuse AdGuard while running by changing the protocol suddenly from IPv4 to IPv6 (or from IPv6 to IPv4). If anybody from the developing team wants to test me something further I’m happily willing to help (I already transmitted the debug logs when contacting the AdGuard support for the first time but it seems that didn’t help much).

I do not plan to upgrade to anything unreliable as Cata(lina|stropha) so the **NetworkExtension API** path isn’t something I’m looking forward to.

**Comment by nexttoyou** [comment]
Addendum: the above solution makes quite reliably disappear the bug after a sleep/wake process. Unfortunately, the bug still reappears when changing back and forth from one user account to another one by fast user switching as well as it happens sometimes even without any reason at all. I’m still unable to find any method of tracing down this annoyance.

**Comment by Ribero** [comment]
It would be nice to hear something from ADGuard dev team about this issue 😉

**Comment by planet0** [comment]
Continuing to occur with macOS Catalina 10.15.6 Beta and the latest AdGuard Nightly build.
When resuming from sleep, AdGuard no longer filters traffic.
Manually quitting and reopening AdGuard fixes the issue until the iMac sleeps again.

**Comment by ameshkov** [comment]
> [Update] Network: I use an IPv6/IPv4 network on Ethernet and Wi-Fi.

This may be interesting.

@planet0 are you on Ethernet as well?

**Comment by planet0** [comment]
@ameshkov Yes, with IPv4 only.
I also have it connected via Wi-Fi to the same network, but Ethernet is prioritised in Network Preferences.

**Comment by ameshkov** [comment]
This seems to be an important detail that we were missing all the time.

@zzebrum could you please try reproducing it with an Ethernet connection

**Comment by planet0** [comment]
@ameshkov I wonder if @Ribero uses the same configuration as me by using both an active Ethernet and Wi-Fi connection at the same time? Would that make any difference?

**Comment by Ribero** [comment]
On iMac I use ethernet (Wifi is activated but not connected to any wifi hotspot). 
On MacBook Pro I'm using WiFi only and Ethernet over Thunderbolt on rare occasion. 

My network is IPv4 and IPv6.

The issue tend to occurs many times more when switching users back & forth + sleep/wake

Currently I have only one user logged on my MacBook and did not get the issue since a while.

**Comment by planet0** [comment]
Hmm. Okay, I’ll keep a close eye on my iMac and collect logs. It’s been happening to me a lot lately, so hopefully I’ll be able to provide something soon.

**Comment by Ribero** [comment]
Today, my wife logged on into the iMac with her account (she does not use this mac often) while I was also logged... then ADG stopped working after weeks of no issues (I was the only one user since 2 weeks+).

Only solution : kill all ADG process

ADG is not usable in a multi-user context.

**Comment by nexttoyou** [comment]
Unfortunately, killing and restarting the AdGuard process under whatever circumstances isn’t the solution. I’m using such workarounds now for switching from one user to another (which is quite [complicated](https://stackoverflow.com/a/57232975), actually) as well as for sleeping and waking the Mac and both still do not make AdGuard working reliably.

The conclusion can only be that AdGuard itself doesn’t know if it is filtering or not and if the pipe is broken for whatever reason nothing at all is done to preserve its work. For those of you using Little Snitch a look inside the log file `/Library/Logs/LittleSnitchDaemon.log` (situated on the root level) might explain whatever provokes a service interruption with the difference that AdGuard ceases its function without communicating while Little Snitch continues perfectly.

And there are obviously many reasons for interrupting AdGuard like a [slow wake process](https://forums.macrumors.com/threads/10-13-6-slow-to-wake-from-sleep.2127447/), a changing IP protocol (IPv4 to IPv6 or the other way around) or some other voodoo I never was able to isolate – e. g. it might be related to the fact that I use two network connections at the same time (Ethernet for the main job and Wi-Fi as link-local only) as otherwise AirDrop wouldn’t work. So as a result I’m hugely disappointed of course as more or less every macOS session might contain a bad surprise …

**Comment by ameshkov** [comment]
Worry not, we'll be able to address it now that we understand how to reproduce the issue. It'll take some time, though, because right now we're quite busy with adding network extensions support.

**Comment by nexttoyou** [comment]
Many **many** thanks for all your efforts!

**Comment by zzebrum** [comment]
reproduced with Ethernet connection and switching between users on Mac Os 10.14 and AdGuard 2.4.8.797 Release

**Comment by zzebrum** [comment]
For those of you who are couraged enough here is a new nightly version: https://agrd.io/mac_nightly (~~_just remember, it may be unstable_~~) to check if it helps to solve the problem. We also will check on our side too.

The kernel extension filtering mode was reworked a lot (this is maybe interesting for @nexttoyou) but the most important change - it has a new network extension filtering mode which is also working in Catalina. To switch to network extension go to the Network tab in AdGuard preference and tap on Select mode button there.

**Comment by nexttoyou** [comment]
The phenomenon still seems to be present with 2.5.0.838 nightly (CL-1.7.2). I did a reboot after the installation and also deactivated all my workarounds.

While out of five fast user switchings there was not one single case of silently disabled filtering, unfortunately, the bug recurred once after a wake up from sleep. Like before the ads were showing up without AdGuard itself realising it and it became necessary to manually switch off the already switched off filtering and then switch it on again (or quit and restart the app).

So I deactivated my workaround for the fast user switching for now but kept my SleepWatcher script.
